### ISSUE TITLE
bucket tool : Auto close gap feature

### ISSUE THREAD

**Issue by blurymind** [mypaint/mypaint#296]
Auto Close gap is a feature that is present in a few manga art related painting applications. 
It is a planned feature for Krita.

What it does:
It tells the bucket tool to automatically close gaps with an invisible straight line, before doing the fill operation. That stop paint from spilling outside of an area, without affecting the classic "Tolerance" value.

Here is an example of its use in paintstorm studio:
https://youtu.be/oXUReniLyqw?t=2m4s

example from clip studio paint:
![closegaplogic](https://cloud.githubusercontent.com/assets/6495061/7296145/95895b90-e9c5-11e4-9f97-1a34c7ca34f9.png)
http://www.clipstudio.net/en/functions

> > [Fill] tool that can fill by closing gaps even when gaps and run over of the fill color are present or by surrounding a small unfilled area.

How is it different from the current "Tolerance" value?
- It doesnt affect the fill's line along the border in a global way like the "Tolerance" value.
  It just stops the fill from spilling by closing the area automatically with a straight line stroke before applying the operation. At least that is how I am assuming it achieves the effect. The "Close Gap Threshold"  value determines the maximum distance between pixels that is allowed before doing the close gap operation.

**Comment by odysseywestra** [comment]
Seems to be a duplicate of #150 and #73, or at least they share the same idea.

**Comment by achadwick** [comment]
**Triage note:** needs algorithm, and one that'll work with MyPaint. Review after 1.2 is released, and close as WONTFIX if no algorithm is present.

**Comment by blurymind** [comment]
it is similar, but it does it in a different way

**Comment by blurymind** [comment]
One important logic note to add to this:
https://cloud.githubusercontent.com/assets/6495061/7296145/95895b90-e9c5-11e4-9f97-1a34c7ca34f9.png

If two lines do not overlap and there is a gap between them: Close gap
If two lines overlap: Do not auto close gap. Not sure how they did it.

BTW an alternative approach to this problem is in gmic

**Comment by 0ion9** [comment]
@blurymind 
I would bet that it is based on gap detection. Naturally, gaps that don't exist won't get closed: they are not detected.

I've made a few casual investigations into how this might work. 

A very simple heuristic might be based on tracking prospective fill edges.Given your example, if we mock-apply a normal bucketfill scanline-by-scanline, tracking the angle over time of the left and right edges, then when the angle changes drastically, we should check for gaps (this is usually where the color otherwise would 'flood out' beyond bounds). This limits CPU demand. 
How to actually check for gaps: 
* examine a limited area (probably something like CloseGapThreshold*2 px wide and high) near the point where the possible gap was detected (this is a 2x2 locale. most likely 1 pixel will be solid and the others transparent).
* Consider each pixel as either line (alpha >= 0.5) or blank (alpha < 0.5)
* Iterate over that mask, finding locations where a line pixel has only one neighbouring line pixel. These are line ends. Usually there will be only 2, I haven't worked out a strategy for if there are more, but one of them should by definition be inside the 2x2 area mentioned earlier.
* if there are exactly two, join them (that is, set pixels on an image for the bucketfill tool's use such that there is a line there)
* continue 'filling'.
* we generated the closes, now we can use our generated image to perform our actual fill (with normal bucketfill)

Some special handling for very small areas (where CGT > total area that should be filled) might be required. In particular it would be possible to detect more than 2 line ends.

This ignores the problem of hitting isolated pixels like GMIC interactive-colorize does, for now. Since we need to generating an energy metric anyway (in order to close), this is "just" a matter of providing a suitable base 'energy' image (ie. it's thankfully a completely isolated and optional component)

It should also be noted that the fill may conceivably 'break out' at both sides at once, and thus, both sides must be checked even if a gap was already filled.

**Comment by blurymind** [comment]
While certainly it would be very difficult to create a silver bullet, I think that it is possible to improve the current fill approach. 

@0ion9 those are very interesting observations. How hard would it be to put a test patch out? :)
:+1: 

Btw, I made a wiki at the inkscape forum, those guys also seem interested in improving their bucket fill tool:
http://wiki.inkscape.org/wiki/index.php/A_better_Bucket_Fill_tool_fill

Krita and gmic developers probably already have some insight in a better fill algorithm. The one in gmic is actually based on a public domain research paper. I need to find it again. It is used in Lazy brush for tvpaint:
http://lazy-brush.com/

I was the one who sent the lazybrush research paper to the gmic developer and he actually implemented it quite well! 

So you guys might be able to collaborate. why not get in touch?

**Comment by 0ion9** [comment]
@blurymind The GMIC approach is IMO superior, but it is a different paradigm (fill everything at once). With the GMIC approach, clicking only once is not possible, as you are required at minimum to define a) the 'inside' part (what you want to fill) and b) the 'outside' part (what you don't want to fill).  Guessing a reasonable value for the latter -- well, I haven't succeeded in it so far, so it's not trivial, that's all I can say.

The most that can be used of GMIC approach is the metric it uses to generate the 'energy map'. I would bet though that this isn't good enough to close gaps, because handling gaps in GMIC often requires a fair bit of trial and error for me.

I confess I don't really care about the algorithm I described, because I consider GMIC to work very well for any non-trivial filling problem. I wrote it up mainly because I recognized that there may be a gain in slightly more 'immediate' filling functionality. It's a fairly simple extension to basic bucket fill, but realistically I'm not gonna code it.

Inkscape, IME, seems to use a classical floodfill, so I would guess that it would also be able to use GMIC's energy metric as a basis for floodfilling.

Whether GMIC's energy metric is actually an improvement, absent the 'watershed' based fill process* is another question.

* watershed process is inherently multi color, so if you are doing single color fill, it cannot be done -- unless you find a reliable heuristic to guess the 'outside' point as I noted, bringing the 'color' count up to 2.
